File: frontend/apps/ios/Pantopus/Core/Analytics/PostHogAnalytics.swift
//
//  PostHogAnalytics.swift
//  Pantopus
//
//  Product-analytics vendor (PostHog) sitting behind the `Analytics` facade.
//  This is the iOS half of the cross-platform pair — the Android counterpart
//  is `data/analytics/PostHogAnalytics.kt`. Both send the SAME event names
//  (the closed `AnalyticsEvent` taxonomy) to the SAME vendor.
//
//  Gating: a no-op until `start` finds a non-empty `PostHogAPIKey` in
//  Info.plist (resolved from the xcconfig / .env `POSTHOG_API_KEY`). Dev and
//  CI builds leave the key blank, so nothing is sent there; beta / prod flip
//  it on with no code change — exactly how `SentryDSN` gates crash reporting.
//
//  Privacy: autocapture is OFF (manual events only), no IDFA / advertising
//  identifier is collected (so `NSPrivacyTracking` stays false and no ATT
//  prompt is required), and event properties are PII-scrubbed before send.
//

import Foundation
import Logging
import PostHog

@MainActor
final class PostHogAnalytics {
    static let shared = PostHogAnalytics()

    private let logger = Logger(label: "app.pantopus.ios.Analytics")
    private(set) var isStarted = false

    private init() {}

    /// Call once from `AppDelegate.didFinishLaunching` (via `Analytics.start`).
    /// Safe to call again — no-ops after the first successful start.
    func start(environment: AppEnvironment) {
        guard !isStarted else { return }

        let apiKey = (Bundle.main.object(forInfoDictionaryKey: "PostHogAPIKey") as? String) ?? ""
        guard !apiKey.isEmpty else {
            logger.info("PostHog API key not set — product analytics disabled")
            return
        }

        let host = Self.host
        let config = PostHogConfig(apiKey: apiKey, host: host)
        // Manual, explicit events ONLY. The taxonomy lives in `AnalyticsEvent`;
        // nothing reaches the wire that isn't declared there.
        config.captureScreenViews = false
        config.captureApplicationLifecycleEvents = false
        // Keep anonymous sessions from minting person profiles — we only create
        // a person when `identify` is called with the app user id (no PII).
        config.personProfiles = .identifiedOnly

        PostHogSDK.shared.setup(config)
        isStarted = true
        logger.info("PostHog started", metadata: ["env": .string(environment.target.rawValue), "host": .string(host)])
    }

    func capture(_ name: String, properties: [String: String]) {
        guard isStarted else { return }
        PostHogSDK.shared.capture(name, properties: Self.scrubPII(properties))
    }

    /// Associate subsequent events with the app user id (pseudonymous — never
    /// email / name). Mirrors `Observability.identify` so crash + analytics
    /// share the same id.
    func identify(userId: String) {
        guard isStarted else { return }
        PostHogSDK.shared.identify(userId)
    }

    /// Drop the user association (sign-out). Starts a fresh anonymous id.
    func reset() {
        guard isStarted else { return }
        PostHogSDK.shared.reset()
    }

    private static var host: String {
        let configured = (Bundle.main.object(forInfoDictionaryKey: "PostHogHost") as? String) ?? ""
        return configured.isEmpty ? "https://eu.i.posthog.com" : configured
    }

    // MARK: - PII scrubbing

    /// Defense-in-depth: the closed taxonomy already carries no PII, but we
    /// still redact any property whose key looks personal and any value that
    /// matches an email / phone pattern before it leaves the device. Mirrors
    /// the scrubbing in `Observability` and the Android vendor.
    private static let piiKeys: Set<String> = [
        "email", "emailaddress", "email_address",
        "phone", "phonenumber", "phone_number", "telephone",
        "address", "street", "streetaddress", "street_address",
        "city", "state", "zip", "zipcode", "zip_code", "postalcode", "postal_code",
        "fullname", "name", "firstname", "first_name", "lastname", "last_name",
        "password", "token", "authorization", "auth", "secret"
    ]

    private static let redacted = "[redacted]"

    private static func scrubPII(_ properties: [String: String]) -> [String: String] {
        var scrubbed = properties
        for (key, value) in properties {
            if piiKeys.contains(key.lowercased()) {
                scrubbed[key] = redacted
            } else {
                scrubbed[key] = redactPatterns(in: value)
            }
        }
        return scrubbed
    }

    private static func redactPatterns(in value: String) -> String {
        var result = value
        let patterns = [
            #"[\w.+-]+@[\w-]+\.[\w.-]+"#,
            #"\+?\d[\d\s().-]{7,}"#
        ]
        for pattern in patterns {
            if let regex = try? NSRegularExpression(pattern: pattern) {
                let range = NSRange(result.startIndex..., in: result)
                result = regex.stringByReplacingMatches(in: result, range: range, withTemplate: redacted)
            }
        }
        return result
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/data/analytics/PostHogAnalytics.kt
package app.pantopus.android.data.analytics

import android.content.Context
import com.posthog.PersonProfiles
import com.posthog.PostHog
import com.posthog.android.PostHogAndroid
import com.posthog.android.PostHogAndroidConfig
import timber.log.Timber

/**
 * Vendor seam so [Analytics] (and tests) never import the PostHog SDK
 * directly. The single concrete implementation is [PostHogAnalytics].
 */
interface AnalyticsVendor {
    fun capture(
        name: String,
        properties: Map<String, String>,
    )

    fun identify(userId: String)

    fun reset()
}

/**
 * PostHog-backed [AnalyticsVendor]. This is the Android half of the
 * cross-platform pair — the iOS counterpart is
 * `Core/Analytics/PostHogAnalytics.swift`. Both send the SAME event names
 * (the closed [AnalyticsEvent] taxonomy) to the SAME vendor.
 *
 * Gating: [create] returns `null` when `POSTHOG_API_KEY` is blank, so
 * analytics no-ops in dev / CI exactly like [app.pantopus.android.data.observability.Observability]
 * does without a Sentry DSN. Beta / prod flip it on with no code change.
 *
 * Privacy: autocapture is OFF (manual events only), no advertising id is
 * collected (so no ATT-equivalent is needed), and event properties are
 * PII-scrubbed before send.
 */
class PostHogAnalytics private constructor() : AnalyticsVendor {
    override fun capture(
        name: String,
        properties: Map<String, String>,
    ) {
        PostHog.capture(event = name, properties = scrubPII(properties))
    }

    override fun identify(userId: String) {
        PostHog.identify(distinctId = userId)
    }

    override fun reset() {
        PostHog.reset()
    }

    companion object {
        private const val DEFAULT_HOST = "https://eu.i.posthog.com"

        /**
         * Initialise PostHog and return a vendor, or `null` when [apiKey] is
         * blank (analytics stays disabled). Call once from
         * `PantopusApplication.onCreate`.
         */
        fun create(
            context: Context,
            apiKey: String,
            host: String,
        ): AnalyticsVendor? {
            if (apiKey.isBlank()) {
                Timber.i("PostHog API key not set — product analytics disabled")
                return null
            }
            val resolvedHost = host.ifBlank { DEFAULT_HOST }
            val config =
                PostHogAndroidConfig(apiKey = apiKey, host = resolvedHost).apply {
                    // Manual, explicit events ONLY — no autocapture. The closed
                    // taxonomy in AnalyticsEvent is the single source of names.
                    captureScreenViews = false
                    captureDeepLinks = false
                    captureApplicationLifecycleEvents = false
                    // Anonymous events don't mint person profiles; a person is
                    // only created on identify() with the app user id (no PII).
                    personProfiles = PersonProfiles.IDENTIFIED_ONLY
                }
            PostHogAndroid.setup(context, config)
            Timber.i("PostHog started (host=$resolvedHost)")
            return PostHogAnalytics()
        }

        /**
         * Defense-in-depth: the closed taxonomy already carries no PII, but we
         * still redact any property whose key looks personal and any value that
         * matches an email / phone pattern. Mirrors the iOS vendor + Sentry
         * scrubbing in Observability.
         */
        private val piiKeys =
            setOf(
                "email", "emailaddress", "email_address",
                "phone", "phonenumber", "phone_number", "telephone",
                "address", "street", "streetaddress", "street_address",
                "city", "state", "zip", "zipcode", "zip_code", "postalcode", "postal_code",
                "fullname", "name", "firstname", "first_name", "lastname", "last_name",
                "password", "token", "authorization", "auth", "secret",
            )
        private const val REDACTED = "[redacted]"
        private val emailRegex = Regex("""[\w.+-]+@[\w-]+\.[\w.-]+""")
        private val phoneRegex = Regex("""\+?\d[\d\s().-]{7,}""")

        private fun scrubPII(properties: Map<String, String>): Map<String, String> =
            properties.mapValues { (key, value) ->
                if (piiKeys.contains(key.lowercase())) {
                    REDACTED
                } else {
                    value.replace(emailRegex, REDACTED).replace(phoneRegex, REDACTED)
                }
            }
    }
}
